File: connectors/java/src/integrationTest/java/com/playhouse/connector/advanced/A07_TlsConnectionTests.java
package com.playhouse.connector.advanced;

import com.playhouse.connector.Connector;
import com.playhouse.connector.ConnectorConfig;
import com.playhouse.connector.Packet;
import com.playhouse.connector.support.BaseIntegrationTest;
import com.playhouse.connector.support.TestMessages.*;
import org.junit.jupiter.api.*;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

import static org.assertj.core.api.Assertions.*;

/**
 * A-07: TLS/WSS 연결 테스트
 * <p>
 * TCP+TLS 및 WSS 전송 계층 연결, 인증, 메시지 송수신 테스트.
 * </p>
 */
@DisplayName("A-07: TLS/WSS 연결 테스트")
@Tag("Advanced")
@Tag("TLS")
class A07_TlsConnectionTests extends BaseIntegrationTest {

    @Override
    @BeforeEach
    public void setUp() throws Exception {
        host = System.getenv().getOrDefault("TEST_SERVER_HOST", "127.0.0.1");
        httpPort = Integer.parseInt(System.getenv().getOrDefault("TEST_SERVER_HTTP_PORT", "28080"));
        httpsPort = Integer.parseInt(System.getenv().getOrDefault("TEST_SERVER_HTTPS_PORT", "28443"));
        tcpPort = Integer.parseInt(System.getenv().getOrDefault("TEST_SERVER_TCP_PORT", "28001"));
        tcpTlsPort = Integer.parseInt(System.getenv().getOrDefault("TEST_SERVER_TCP_TLS_PORT", "28002"));

        testServer = new com.playhouse.connector.support.TestServerClient(host, httpPort);
        stageInfo = testServer.createTestStage();

        connector = new Connector();
    }

    @Test
    @DisplayName("A-07-01: TCP+TLS로 서버에 연결할 수 있다")
    void tcpTlsConnectionSuccess() throws Exception {
        connector.init(ConnectorConfig.builder()
                .useSsl(true)
                .skipServerCertificateValidation(true)
                .requestTimeoutMs(5000)
                .heartbeatIntervalMs(10000)
                .build());

        connector.setStageId(stageInfo.getStageId());
        CompletableFuture<Void> connectFuture = connector.connectAsync(host, tcpTlsPort);
        connectFuture.get(5, TimeUnit.SECONDS);

        assertThat(connector.isConnected()).isTrue();

        AuthenticateRequest authRequest = new AuthenticateRequest("tls-user-1", "valid_token");
        Packet authPacket = Packet.builder("AuthenticateRequest")
                .payload(authRequest.toByteArray())
                .build();

        Packet authReplyPacket = connector.requestAsync(authPacket)
                .get(5, TimeUnit.SECONDS);
        AuthenticateReply authReply = AuthenticateReply.parseFrom(authReplyPacket.getPayload());
        assertThat(authReply.success).isTrue();

        EchoRequest echoRequest = new EchoRequest("Hello TLS", 1);
        Packet echoPacket = Packet.builder("EchoRequest")
                .payload(echoRequest.toByteArray())
                .build();

        Packet echoReplyPacket = connector.requestAsync(echoPacket)
                .get(5, TimeUnit.SECONDS);
        EchoReply echoReply = EchoReply.parseFrom(echoReplyPacket.getPayload());
        assertThat(echoReply.content).isEqualTo("Hello TLS");
    }

    @Test
    @DisplayName("A-07-02: WSS로 서버에 연결할 수 있다")
    void wssConnectionSuccess() throws Exception {
        connector.init(ConnectorConfig.builder()
                .useWebsocket(true)
                .useSsl(true)
                .webSocketPath("/ws")
                .skipServerCertificateValidation(true)
                .requestTimeoutMs(5000)
                .heartbeatIntervalMs(10000)
                .build());

        connector.setStageId(stageInfo.getStageId());
        CompletableFuture<Void> connectFuture = connector.connectAsync(host, httpsPort);
        connectFuture.get(5, TimeUnit.SECONDS);

        assertThat(connector.isConnected()).isTrue();

        AuthenticateRequest authRequest = new AuthenticateRequest("wss-user-1", "valid_token");
        Packet authPacket = Packet.builder("AuthenticateRequest")
                .payload(authRequest.toByteArray())
                .build();

        Packet authReplyPacket = connector.requestAsync(authPacket)
                .get(5, TimeUnit.SECONDS);
        AuthenticateReply authReply = AuthenticateReply.parseFrom(authReplyPacket.getPayload());
        assertThat(authReply.success).isTrue();

        EchoRequest echoRequest = new EchoRequest("Hello WSS", 2);
        Packet echoPacket = Packet.builder("EchoRequest")
                .payload(echoRequest.toByteArray())
                .build();

        Packet echoReplyPacket = connector.requestAsync(echoPacket)
                .get(5, TimeUnit.SECONDS);
        EchoReply echoReply = EchoReply.parseFrom(echoReplyPacket.getPayload());
        assertThat(echoReply.content).isEqualTo("Hello WSS");
    }
}

File: connectors/csharp/tests/PlayHouse.Connector.IntegrationTests/AdvancedTests/A07_TlsConnectionTests.cs
using PlayHouse.Connector.Protocol;
using PlayHouse.TestServer.Proto;
using Xunit;

namespace PlayHouse.Connector.IntegrationTests.AdvancedTests;

/// <summary>
/// A-07: TLS/WSS 연결 테스트
/// </summary>
/// <remarks>
/// TCP+TLS 및 WSS 전송 계층 연결, 인증, Echo 동작을 검증합니다.
/// </remarks>
[Trait("Category", "Advanced")]
[Trait("Transport", "TLS")]
public class A07_TlsConnectionTests : IClassFixture<TestServerFixture>, IAsyncLifetime
{
    private readonly TestServerFixture _testServer;
    private Connector? _connector;
    private CreateStageResponse? _stageInfo;

    public A07_TlsConnectionTests(TestServerFixture testServer)
    {
        _testServer = testServer;
    }

    public async Task InitializeAsync()
    {
        _connector = new Connector();
        _stageInfo = await _testServer.CreateTestStageAsync();
    }

    public async Task DisposeAsync()
    {
        if (_connector != null)
        {
            if (_connector.IsConnected())
            {
                _connector.Disconnect();
                await Task.Delay(100);
            }
            await _connector.DisposeAsync();
        }
    }

    [Fact(DisplayName = "A-07-01: TCP+TLS로 서버에 연결할 수 있다")]
    public async Task TcpTls_Connection_Success()
    {
        _connector!.Init(new ConnectorConfig
        {
            UseWebsocket = false,
            UseSsl = true,
            SkipServerCertificateValidation = true,
            RequestTimeoutMs = 5000,
            HeartBeatIntervalMs = 10000
        });

        var connected = await _connector.ConnectAsync(
            _testServer.Host,
            _testServer.TcpTlsPort,
            _stageInfo!.StageId,
            _stageInfo.StageType
        );

        Assert.True(connected);
        Assert.True(_connector.IsConnected());

        var authRequest = new AuthenticateRequest { UserId = "tls-user-1", Token = "valid_token" };
        using var authPacket = new Packet(authRequest);
        var authReplyPacket = await _connector.AuthenticateAsync(authPacket);
        var authReply = AuthenticateReply.Parser.ParseFrom(authReplyPacket.Payload.DataSpan.ToArray());
        Assert.True(authReply.Success);

        var echoRequest = new EchoRequest { Content = "Hello TLS", Sequence = 1 };
        using var echoPacket = new Packet(echoRequest);
        var echoReplyPacket = await _connector.RequestAsync(echoPacket);
        var echoReply = EchoReply.Parser.ParseFrom(echoReplyPacket.Payload.DataSpan.ToArray());
        Assert.Equal("Hello TLS", echoReply.Content);
    }

    [Fact(DisplayName = "A-07-02: WSS로 서버에 연결할 수 있다")]
    public async Task Wss_Connection_Success()
    {
        _connector!.Init(new ConnectorConfig
        {
            UseWebsocket = true,
            UseSsl = true,
            SkipServerCertificateValidation = true,
            WebSocketPath = "/ws",
            RequestTimeoutMs = 5000,
            HeartBeatIntervalMs = 10000
        });

        var connected = await _connector.ConnectAsync(
            _testServer.Host,
            _testServer.HttpsPort,
            _stageInfo!.StageId,
            _stageInfo.StageType
        );

        Assert.True(connected);
        Assert.True(_connector.IsConnected());

        var authRequest = new AuthenticateRequest { UserId = "wss-user-1", Token = "valid_token" };
        using var authPacket = new Packet(authRequest);
        var authReplyPacket = await _connector.AuthenticateAsync(authPacket);
        var authReply = AuthenticateReply.Parser.ParseFrom(authReplyPacket.Payload.DataSpan.ToArray());
        Assert.True(authReply.Success);

        var echoRequest = new EchoRequest { Content = "Hello WSS", Sequence = 2 };
        using var echoPacket = new Packet(echoRequest);
        var echoReplyPacket = await _connector.RequestAsync(echoPacket);
        var echoReply = EchoReply.Parser.ParseFrom(echoReplyPacket.Payload.DataSpan.ToArray());
        Assert.Equal("Hello WSS", echoReply.Content);
    }
}
